artemis-core-client/src/main/java/org/apache/activemq/artemis/core/client/impl/ClientProducerImpl.java

```diff
@@ -138,24 +135,19 @@ public void send(SimpleString address1,
                     Message message,
                     SendAcknowledgementHandler handler) throws ActiveMQException {
       checkClosed();
-      boolean confirmationWindowEnabled = session.isConfirmationWindowEnabled();
-      if (confirmationWindowEnabled) {
-         doSend(address1, message, handler);
-      } else {
-         doSend(address1, message, null);
-         if (handler != null) {
-            if (logger.isDebugEnabled()) {
-               logger.debug("Handler was used on producing messages towards address " + (address1 == null ? null : address1.toString()) + " however there is no confirmationWindowEnabled");
-            }
+      if (handler != null) {
+         handler = new SendAcknowledgementHandlerWrapper(handler);
```

Comment:
this will create a new object for every message send called when no handlers in place?

that's a heavy change, no?

Reply:
No, it won't create a new object for every message send. It will only create a wrapper for the handler if a handler is set when a message is sent.